### PR TITLE
tests: py2/3 unified, using pytest, automated on Travis

### PR DESCRIPTION
**Pull request by temoto**
Plan:
- merge py2/3 tests (done here)
- merge py2/3 code while keeping merged tests pass
- finish before next Ubuntu LTS

Related: #25 #42 #53

### CODE REVIEW

File: python2/httplib2/__init__.py [hunk 1]
```diff
@@ -962,7 +977,8 @@ def connect(self):
                 continue
             break
         if not self.sock:
-            raise socket.error(msg)
+            raise socket.error, msg
```
Comment (cdent) [inline python2/httplib2/__init__.py]: This seems like a step backwards to syntax that is invalid for py3? I recognize that we're in the python2 tree here, but it's not clear why this change was made?

Comment (temoto) [reply python2/httplib2/__init__.py]: Yes, I am guilty of surprise here, should've left a comment or put it in separate commit.

I didn't dig deep, so don't fully understand how it works, but having it raise `socket.error()` instance makes regression to tests that fail socket connect. My guess it's related to shadowing `msg` in earlier try/except branch, but that's only a guess, possibly wrong one.

See https://github.com/httplib2/httplib2/pull/81

**Comment by cclauss** [comment]
I am still seeing the following... That __new_request__ is an undefined name and that __httplib2__ is an undefined name inside of __httplib2__.

flake8 testing of https://github.com/httplib2/httplib2 on Python 2.7.14

$ __flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__
```
./python2/httplib2test.py:1239:20: F821 undefined name 'new_request'
            return new_request(*args, **kwargs)
                   ^
./python2/httplib2/__init__.py:890:24: F821 undefined name 'httplib2'
        bypass_hosts = httplib2.AllHosts
                       ^
./python3/httplib2test.py:1182:20: F821 undefined name 'new_request'
            return new_request(*args, **kwargs)
                   ^
3     F821 undefined name 'httplib2'
```

**Comment by temoto** [comment]
@cclauss right, consider python{2,3}/*test* deprecated, they will be deleted.

I started with deleting parts of old tests along with rewriting but then decided to restore original test files in case someone has a grasp on old tests and can catch important difference without going to git history.

**Comment by temoto** [comment]
@cclauss this one is a legitimate bug though, it was fixed, I guess you are looking at older version. `git fetch && git checkout tests && git pull --rebase`

**Comment by temoto** [comment]
Your attention is greatly appreciated.

**Comment by cclauss** [comment]
Thanks for the explanations.  It makes sense to me now.  You are correct...  I was running flake8 on __master__ not on __tests__.

**Comment by cclauss** [comment]
The __tests__ branch looks good on the key flake8 items ([--select=E901,E999,F821,F822,F823](https://github.com/cclauss/flake8-for-httplib2/blob/master/.travis.yml#L23))...
* https://travis-ci.org/cclauss/flake8-for-httplib2

**Comment by temoto** [comment]
@cclauss it also executes flake8 ignore=E261,W503 as part of Travis build.
https://travis-ci.org/httplib2/httplib2/builds/343220554

**Comment by temoto** [comment]
@stinky2nine ping

**Comment by temoto** [comment]
I guess it was cooking long enough and is ready by now.

**Comment by cclauss** [comment]
Great work!  Will there be a push to PyPI?  It would be nice to see httplib2 turn green the WOS.... https://python3wos.appspot.com

**Comment by temoto** [comment]
Of course I will publish pypi release, though not decided yet what exactly
should land there.
Probably digicert certificate should be included.

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+cache:
+  apt: true
+  ccache: true
+  pip: true
+  directories:
+    - $HOME/.cache
+addons:
+  apt_packages:
+    # - libssl-dev
```
Comment (stinky2nine) [inline .travis.yml]: remove if not needed

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
+sudo: false
+language: python
+
+matrix:
+  fast_finish: true
+  include:
+    - {python: 2.7, env: test_group=pep8}
+    - {python: 3.6, env: test_group=pep8}
+    - {python: 2.7}
+    - {python: 3.3}
```
Comment (stinky2nine) [inline .travis.yml]: do we care about 3.3 which is already EOL last September?

**Comment by stinky2nine** [comment]
@temoto Sorry, I just came back from a long vacation. LGTM overall, had two comments.

**Comment by stinky2nine** [comment]
Wrt to certificates, instead of us making the decision, could we just pull from trustworthy source like Chrome or Firefox and use their official list?

**Comment by temoto** [comment]
I'm checking that digicert patch by hand (against Mozilla list) last time and then I'll present pypi dependency on `certifi` (which is Mozilla list again).

File: .travis.yml (hunk 2 above)
Comment (temoto) [reply .travis.yml]: Thanks.

File: .travis.yml (hunk 3 above)
Comment (temoto) [reply .travis.yml]: Basically, the only 3.3 specific part is `SSLContext.check_hostname/match_hostname`
as we've found out in incredible work #90 

I think it's cheap enough to leave for a while.

 https://github.com/httplib2/httplib2/pull/90/files#diff-72a22fe2efc9771bbe023ba1ffa07173R146

Comment (stinky2nine) [reply .travis.yml]: SGTM
